### PR TITLE
Feat: Basic Launchbox contracts

### PR DESCRIPTION
**Pull request by codebuster22**
What I have done:
- Add ERC20 and Factory contracts for deploying contracts with exchange
- Add a basic LaunchboxExchange contracts
- Added 100% test coverage for ERC20 and Factory contracts

To be done in next PR:
- Add the logic for conversion of tokens and calculating price.
- Add the logic of fetching and calculating actual market cap of token in USD.

### CODE REVIEW

File: src/launchbox/exchange/LaunchboxExchange.sol [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+contract LaunchboxExchange {
+    IERC20 public token;
+    uint256 public maxSupply;
+    uint256 public marketCapThreshold;
+    uint256 public launchboxErc20Balance;
+    uint256 public ethBalance;
+
+    bool public saleActive = false;
+
+    IUniswapV2Router02 public uniswapRouter;
```
Comment (njokuScript) [inline src/launchbox/exchange/LaunchboxExchange.sol]: Can we make use of Aerodrome instead?

File: src/launchbox/exchange/LaunchboxExchange.sol [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        launchboxErc20Balance += tokenAmount;
+        require(token.transferFrom(msg.sender, address(this), tokenAmount), "Transfer failed");
+        if(address(this).balance < ethToReturn) {
+            revert NotEnoughETH();
+        }
+
+        payable(msg.sender).transfer(ethToReturn);
+    }
+
+    function endBonding() internal {
```
Comment (njokuScript) [inline src/launchbox/exchange/LaunchboxExchange.sol:92]: I see this function puts the liquidity into the pool.

Can we put only 50% of the ETH? and then progressively deploy the Liquidity over a period of 1 month?

File: src/launchbox/token/LaunchboxFactory.sol [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        launchboxExchangeImplementation = _launchboxExchangeImplementation;
+        uniswapRouter = _uniswapRouter;
+    }
+
+    function deployToken(
+        string memory name,
+        string memory symbol,
+        uint256 maxSupply,
+        string memory metadataURI,
+        uint256 marketCapThreshold
```
Comment (njokuScript) [inline src/launchbox/token/LaunchboxFactory.sol]: `marketCapThreshold` should not come from the frontend. Ideally, it should be one of the admin parameters to be set in the Exchange contract

File: src/launchbox/token/LaunchboxERC20.sol [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            revert MetadataEmpty();
+        }
+        if(_maxSupply == 0) {
+            revert CannotSellZeroTokens();
+        }
+
+        metadataURI = _metadataURI;
+
+        launchboxExchange = payable(Clones.clone(_launchboxExchangeImplementation));
+        _mint(launchboxExchange, _maxSupply);
```
Comment (njokuScript) [inline src/launchbox/token/LaunchboxERC20.sol]: Ideally, we would want to calculate the fees and pass that into the `intitalize` method. Fee should be set by admin in the factory contract. Fee will be a % of the token supply that will be sent to a Launchbox managed address. see here: https://github.com/metastable-labs/contracts/blob/launchbox-erc20/src/launchbox/token/LaunchboxERC20Factory.sol#L69-L71

File: src/launchbox/token/LaunchboxFactory.sol [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    address public tokenImplementation;
+    address public launchboxExchangeImplementation;
+    address public uniswapRouter;
+    
+    error EmptyTokenImplementation();
+    error EmptyLaunchboxExchangeImplementation();
+    error EmptyUniswapRouter();
+
+    constructor(address _tokenImplementation, address _launchboxExchangeImplementation, address _uniswapRouter) {
```
Comment (njokuScript) [inline src/launchbox/token/LaunchboxFactory.sol]: We also need to implement first buy. so after the contracts gets created, the deployer should be a certain portion of the token.

File: src/launchbox/token/LaunchboxERC20.sol (hunk 4 above)
Comment (njokuScript) [reply src/launchbox/token/LaunchboxERC20.sol]: Screen Shot 2024-06-13 at 7 40 42 AM
how I'm thinking about distributing the supply

File: src/launchbox/exchange/LaunchboxExchange.sol (hunk 1 above)
Comment (codebuster22) [reply src/launchbox/exchange/LaunchboxExchange.sol]: Cool.

File: src/launchbox/exchange/LaunchboxExchange.sol (hunk 2 above)
Comment (codebuster22) [reply src/launchbox/exchange/LaunchboxExchange.sol:92]: Someone will have to keep calling this function. The smart contract won't be able to put the liquidity progressively automatically.

File: src/launchbox/token/LaunchboxFactory.sol (hunk 5 above)
Comment (codebuster22) [reply src/launchbox/token/LaunchboxFactory.sol]: cool, but we shouldn't make it compulsory right?

File: src/launchbox/token/LaunchboxFactory.sol (hunk 3 above)
Comment (codebuster22) [reply src/launchbox/token/LaunchboxFactory.sol]: I thought of keeping it open ended. Also, saw you have written the contracts on another branch. Didn't saw a PR for that so I wrote them too. I'll remove the factory and erc20.

File: src/launchbox/exchange/LaunchboxExchange.sol (hunk 2 above)
Comment (njokuScript) [reply src/launchbox/exchange/LaunchboxExchange.sol:92]: Yeah, we can setup a script for that

File: src/launchbox/token/LaunchboxFactory.sol (hunk 3 above)
Comment (njokuScript) [reply src/launchbox/token/LaunchboxFactory.sol]: Nah, don't remove yours. let's keep yours and just make the changes. Let that open ended decision come from us not the users

File: src/launchbox/token/LaunchboxFactory.sol (hunk 5 above)
Comment (njokuScript) [reply src/launchbox/token/LaunchboxFactory.sol]: yeah

Comment (njokuScript) [reply src/launchbox/token/LaunchboxFactory.sol]: I think we should take it out entirely. extra complexity

File: src/launchbox/exchange/LaunchboxExchange.sol [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        maxSupply = _maxSupply;
+        marketCapThreshold = _marketCapThreshold;
+        launchboxErc20Balance = token.balanceOf(address(this));
+        launchboxErc20BalanceReceived = launchboxErc20Balance;
+        reserveRatio = 100_000; // 10%
+        ethBalance = address(this).balance;
+        saleActive = true;
+    }
+
+    function buyTokens() public payable {
```
Comment (iamnotstatic) [inline src/launchbox/exchange/LaunchboxExchange.sol]: Is possible to emit a Buy event here ?

File: src/launchbox/exchange/LaunchboxExchange.sol [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    function calculateMarketCap() external view returns (uint256) {
+        return _calculateMarketCap();
+    }
+
+    function getTokenPriceinETH() external view returns (uint256 ethAmount) {
+        return _convertToSellTokens(1 * 1e18);
+    }
+
+    function sellTokens(uint256 tokenAmount) public {
```
Comment (iamnotstatic) [inline src/launchbox/exchange/LaunchboxExchange.sol:77]: Same with sell event ?

**Review by njokuScript — APPROVED** [review]

**Pull request merged**